File: swift/APITest/APITest/APITest.swift
//
//  Copyright © 2016 Zello. All rights reserved.
//

import Foundation

open class APITest {
  
  var api: ZelloAPI!
  
  public convenience init(host: String, apiKey: String, username: String, password: String) {
    self.init()
    
    api = ZelloAPI(host: host, apiKey: apiKey)
    authenticate(username, password: password)
  }
  
  fileprivate func authenticate(_ username: String, password: String) {
    api.authenticate(username, password: password) { [weak self] (success, response, error) in
      print("authenticate: " + String(success))

      if success {
        self?.startTesting()
      } else {
        print("Input the correct credentials for your network in ViewController.swift")
      }
    }
  }
  
  fileprivate func startTesting() {
    api.getUsers { [weak self] (success, response, error) in
      print("getUsers: " + String(success))

      self?.api.getChannels(completionHandler: { (success, response, error) in
        print("getChannels: " + String(success))
        if success {
          print(response)
        }
        
        // Add or update user
        var userDictionary = [String : String]()
        userDictionary["name"] = "zelloapi_test"
        userDictionary["password"] = "test".MD5()
        userDictionary["email"] = "support@zello.com"
        userDictionary["full_name"] = "API Test 'На здоровье'" // UTF-8 is fully supported
        self?.api.saveUser(userDictionary, completionHandler: { (success, response, error) in
          print("saveUser: " + String(success))

          // List users again -- look the new user is there
          self?.api.getUsers(completionHandler: { (success, response, error) in
            print("getUsers: " + String(success))

            if success {
              print(response)
            }
            
            self?.continueTesting()
          })
        })
      })
    }
  }
  
  fileprivate func continueTesting() {
    // Add channel
    api.addChannel("Test channel") { [weak self] (success, response, error) in
      print("addChannel: " + String(success))

      // Add user to a channel
      let users = ["zelloapi_test"]
      self?.api.addToChannel("Test channel", users: users, completionHandler: { (success, response, error) in
        print("addToChannel: " + String(success))

        // List channels again
        self?.api.getChannels(completionHandler: { (success, response, error) in
          print("getChannels: " + String(success))

          if success {
            print(response)
          }
          
          // Create channel role
          
          var channelRoleDictionary = [String : AnyObject]()
          channelRoleDictionary["listen_only"] = false as AnyObject?
          channelRoleDictionary["no_disconnect"] = true as AnyObject?
          channelRoleDictionary["allow_alerts"] = false as AnyObject?
          var toArray: [AnyObject] = []
          channelRoleDictionary["to"] = toArray as AnyObject?
          
          self?.api.saveChannelRole("Test channel", roleName: "Dispatcher", settings: channelRoleDictionary, completionHandler: { (success, response, error) in
            print("saveChannelRole: " + String(success))

            channelRoleDictionary = [String : AnyObject]()
            channelRoleDictionary["listen_only"] = false as AnyObject?
            channelRoleDictionary["no_disconnect"] = false as AnyObject?
            channelRoleDictionary["allow_alerts"] = true as AnyObject?
            toArray = ["Dispatcher" as AnyObject]
            channelRoleDictionary["to"] = toArray as AnyObject?
            
            self?.api.saveChannelRole("Test channel", roleName: "Driver", settings: channelRoleDictionary, completionHandler: { (success, response, error) in
              print("saveChannelRole: " + String(success))

              // List channel roles
              self?.api.getChannelsRoles("Test channel", completionHandler: { (success, response, error) in
                print("getChannelsRoles: " + String(success))

                if success {
                  print(response)
                }
                
                self?.cleanUp()
              })
            })
          })
        })
      })
    }
  }
  
  fileprivate func cleanUp() {
    // Remove the channel
    let channelNames = ["Test channel"]
    api.deleteChannels(channelNames) { [weak self] (success, response, error) in
      print("deleteChannels: " + String(success))

      // Delete the user we just added
      let users = ["zelloapi_test"]
      self?.api.deleteUsers(users, completionHandler: { (success, response, error) in
        print("deleteUsers: " + String(success))

        // List users one last time -- the new user is gone
        self?.api.getUsers(completionHandler: { (success, response, error) in
          print("getUsers: " + String(success))

          if success{
            print(response)
          }
        })
      })
    }
  }
  
}

private extension String {
  
  /// Calculates the MD5 hash of a string.
  func MD5() -> String {
    guard let str = cString(using: String.Encoding.utf8) else {
      return ""
    }
    
    let strLen = CC_LONG(lengthOfBytes(using: String.Encoding.utf8))
    let digestLen = Int(CC_MD5_DIGEST_LENGTH)
    let result = UnsafeMutablePointer<CUnsignedChar>.allocate(capacity: digestLen)
    
    CC_MD5(str, strLen, result)
    
    let hash = NSMutableString()
    for i in 0..<digestLen {
      hash.appendFormat("%02x", result[i])
    }
    
    result.deallocate(capacity: digestLen)
    
    return String(format: hash as String)
  }
}

File: csharp/APITest/APITest/APITest.cs
//
//  Copyright © 2016 Zello. All rights reserved.
//

using System;
using System.Collections.Generic;
using System.Collections;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace APITest
{
    public class APITest
    {
        readonly ZelloAPI api;

        public APITest(string host, string apiKey)
        {
            api = new ZelloAPI(host, apiKey);
        }

        public async Task startTesting(string username, string password)
        {
            await authenticate(username, password);
        }

        async Task authenticate(string username, string password)
        {
            ZelloAPIResult result = null;
            try
            {
                result = await api.Authenticate(username, password);
            }
            catch (System.UriFormatException)
            {
            }

            if (result != null)
            {
                Console.WriteLine("Authenticate: " + result.Success);

                if (result.Success)
                {
                    await callOtherMethods();
                }
                else
                {
                    Console.WriteLine("Input the correct credentials for your network in Program.cs");
                }
            }
            else
            {
                Console.WriteLine("Input the correct credentials for your network in Program.cs");
            }
        }

        async Task callOtherMethods()
        {
            ZelloAPIResult result = await api.GetUsers(null, false, null, null, null);
            Console.WriteLine("GetUsers: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["users"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }

            result = await api.GetChannels(null, null, null);
            Console.WriteLine("GetChannels: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["channels"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }

            // Add or update user
            var userDictionary = new Dictionary<string, string>();
            userDictionary.Add("name", "zelloapi_test");
            userDictionary.Add("password", MD5Hash("test"));
            userDictionary.Add("email", "support@zello.com");
            userDictionary.Add("full_name", "API Test 'На здоровье'"); // UTF-8 is fully supported 
            result = await api.SaveUser(userDictionary);
            Console.WriteLine("SaveUser: " + result.Success);

            // List users again -- look the new user is there
            result = await api.GetUsers(null, false, null, null, null);
            Console.WriteLine("GetUsers: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["users"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }

            // Add channel
            result = await api.AddChannel("Test channel", null, null);
            Console.WriteLine("AddChannel: " + result.Success);

            // Add user to a channel
            var users = new ArrayList();
            users.Add("zelloapi_test");
            result = await api.AddToChannel("Test channel", users);
            Console.WriteLine("AddToChannel: " + result.Success);

            // List channels again
            result = await api.GetChannels(null, null, null);
            Console.WriteLine("GetChannels: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["channels"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }

            // Create channel role

            var channelRoleDictionary = new Dictionary<string, object>();
            channelRoleDictionary.Add("listen_only", false);
            channelRoleDictionary.Add("no_disconnect", true);
            channelRoleDictionary.Add("allow_alerts", false);
            var toArray = new string[0];
            channelRoleDictionary.Add("to", toArray);
            result = await api.SaveChannelRole("Test channel", "Dispatcher", channelRoleDictionary);
            Console.WriteLine("SaveChannelRole: " + result.Success);

            channelRoleDictionary = new Dictionary<string, object>();
            channelRoleDictionary.Add("listen_only", false);
            channelRoleDictionary.Add("no_disconnect", false);
            channelRoleDictionary.Add("allow_alerts", true);
            toArray = new string[] { "Dispatcher" };
            channelRoleDictionary.Add("to", toArray);
            result = await api.SaveChannelRole("Test channel", "Driver", channelRoleDictionary);
            Console.WriteLine("SaveChannelRole: " + result.Success);

            // List channel roles
            result = await api.GetChannelsRoles("Test channel");
            Console.WriteLine("GetChannelsRoles: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["roles"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }

            // Remove the channel
            var channelNames = new ArrayList();
            channelNames.Add("Test channel");
            result = await api.DeleteChannels(channelNames);

            // Delete the user we just added
            users = new ArrayList();
            users.Add("zelloapi_test");
            result = await api.DeleteUsers(users);

            // List users one last time -- the new user is gone
            result = await api.GetUsers(null, false, null, null, null);
            Console.WriteLine("GetUsers: " + result.Success);
            if (result.Success)
            {
                Object[] arr = (Object[])result.Response["users"];
                foreach (Object obj in arr)
                {
                    dictionaryOut((Dictionary<string, object>)obj);
                }
            }
        }

        string MD5Hash(string input)
        {
            // Convert the input string to a byte array and compute the hash.
            byte[] data = MD5.Create().ComputeHash(Encoding.UTF8.GetBytes(input));

            // Create a new StringBuilder to collect the bytes
            // and create a string.
            var sBuilder = new StringBuilder();

            // Loop through each byte of the hashed data 
            // and format each one as a hexadecimal string.
            for (int i = 0; i < data.Length; i++)
            {
                sBuilder.Append(data[i].ToString("x2"));
            }

            // Return the hexadecimal string.
            return sBuilder.ToString();
        }

        void dictionaryOut(Dictionary<string, object> dictionary)
        {
            foreach (KeyValuePair<string, object> temp in dictionary)
            {
                Console.WriteLine(temp.Key + " : " + temp.Value.ToString());
            }

            Console.WriteLine();
        }
    }
}
